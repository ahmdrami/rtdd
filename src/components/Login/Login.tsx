import React, { useReducer } from 'react'

interface ILogin {
  resolved: boolean
  loading: boolean
  error: string
}

export function Login() {
  const [state, setState] = useReducer(
    (s: ILogin, a: any) => ({ ...s, ...a }),
    {
      resolved: false,
      loading: false,
      error: null
    }
  )
  function handleSubmit(event: any) {
    event.preventDefault()
    const { username, password } = event.target.elements
    setState({ resolved: false, loading: true, error: null })

    fetch('/api/login', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        username: username.value,
        password: password.value
      })
    })
      .then(r => r.json())
      .then(() => setState({ loading: false, resolved: true, error: null }))
      .catch(e => setState({ loading: false, resolved: false, error: 'error' }))
  }
  return (
    <div>
      <form onSubmit={handleSubmit}>
        <div>
          <label htmlFor="username">Username</label>
          <input id="username" />
        </div>
        <div>
          <label htmlFor="password">Password</label>
          <input id="password" />
        </div>
        <button type="submit">Submit</button>
      </form>
      {state.error && <div role="alert">error</div>}
      {state.resolved && <div role="alert">Signed in</div>}
    </div>
  )
}
