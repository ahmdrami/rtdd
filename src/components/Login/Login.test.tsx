import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import { Login } from './Login'

describe('Login component', () => {
  test('Allow users to login successfully', async () => {
    const fakeUserResponse = { token: 'fake_user_token ' }

    jest.spyOn(window, 'fetch').mockImplementationOnce((): any =>
      Promise.resolve({
        json: () => Promise.resolve(fakeUserResponse)
      })
    )
    render(<Login />)

    fireEvent.change(screen.getByLabelText(/username/i), { value: 'John' })

    fireEvent.change(screen.getByLabelText(/password/i), { value: 'doe' })

    fireEvent.click(screen.getByText(/submit/i))

    // just like a manual tester, we'll instruct our test to wait for the alert
    // to show up before continuing with our assertions.
    const alert = await screen.findByRole('alert')
    // .toHaveTextContent() comes from jest-dom's assertions
    // otherwise you could use expect(alert.textContent).toMatch(/congrats/i)
    // but jest-dom will give you better error messages which is why it's recommended
    expect(alert).toHaveTextContent(/signed in/i)
  })
})
