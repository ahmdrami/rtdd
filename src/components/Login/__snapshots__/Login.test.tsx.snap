// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component Snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <div>
        <label
          for="username"
        >
          Username
        </label>
        <input
          id="username"
        />
      </div>
      <div>
        <label
          for="password"
        >
          Password
        </label>
        <input
          id="password"
        />
      </div>
      <button
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
